### DISCUSSION TITLE
Github autolinks incorrect when merging histories (many repos to monorepo)

### DISCUSSION THREAD

**Discussion by chrisgbaker** [community/community#47834] [Pull Requests]
### Select Topic Area

Bug

### Body

Hello!
I do not know if this should go under Pull Requests, or if it's technically a bug or a feature request, so feel free to move is there is a more appropriate area.

Currently my team is working across 350+ microservice repos, and are beginning research of what a monorepo would look like for us. In doing so, we are attempting to merge these microservices into a current repo, retaining all relevant history, either by using subtrees, or by creating a temp commit in the microservice, removing the submodule, and fetching that temp submodule commit as a separate origin, merging it into the main repo.

Both of these methods appear to work in retaining history/commits, however when Github creates an autolink in the UI, the autolink is simply to something like `#50`; the issue here is that the link will be for the _current_ repo, not the submodule's repo (`previous_submodule_a#50`). 

Is there any way to import multiple repos into a single repo maintaining relevant history and also have these autolinks correct? It looks like this is the behavior if you fork a repo in github, as far as preserving the linking and previous repo's history.

**Comment by Dvir-Cohen1** [comment]
Hello!

When you merge multiple microservices into a single repository, you are essentially consolidating the commit histories of the individual microservices into one repository. As a result, the commit hashes and references to issues or pull requests will change, as they are now part of a new repository. This can cause autolinks to break and point to the wrong location.

To preserve the history and autolinks, you may want to consider using Git's subtree merge strategy or a Git submodule. However, as you have mentioned, these methods do not seem to be working for you.

If you are looking for a way to import multiple repositories into a single repository while maintaining the commit history and autolinks, you may want to consider using Git's graft feature or a third-party tool such as git-filter-repo or git-subtree-split.

The git-filter-repo tool can help you split a repository into multiple sub-repositories while preserving the commit history and references to issues or pull requests. Similarly, the git-subtree-split tool can help you split a repository into multiple sub-repositories and maintain the history and links.

I hope this helps!

**Reply by chrisgbaker** [reply]
Thanks for the info! I think the tool you're suggesting however does the opposite of what I am looking to do. From the descriptions, it seems `git-filter-repo` is for taking a large repo and breaking it down into smaller subfolders/submodules.

We currently have a "main" repo, which has X many submodules. We **want** to have a "main" repo which _includes_ all of those submodules as regular folders, retaining the previous submodules history. This works by automating moving the files, removing the submodule, then using `git merge –allow-unrelated-histories` after fetching the submodules origin into a folder (say `./submoduleA`), but auto-generated links from former history are incorrect.

In looking at `git-filter-repo` though, it looks like you may be able to simply rewrite history messages without moving files? In that case it may do the trick, turning `(#50)` into `(submoduleA#50)` for instance.

**Edit/Followup:** It looks like you _can_ simply rewrite history without moving files, which almost solved the problem. Though the end goal I now do not know if what I am attempting will be possible. With a modified commit message the sha will change, which we want to stay the same for historical purposes.

**Reply by Dvir-Cohen1** [reply]
You're correct that if you use git-filter-repo to modify the commit messages in your submodules, the SHA hashes of the commits will change, since the SHA hash is based on the contents of the commit, including its commit message. If you want to preserve the original SHA hashes of the commits, you'll need to find another way to modify the autolinks without modifying the commit messages.

One possibility is to use git replace to create a mapping between the original commit in the submodule and a new commit in the main repository. This would allow you to modify the autolinks to point to the new commits in the main repository without changing the SHA hashes of the original commits.

Here's a rough outline of the steps involved:

1. Clone the main repository and all the submodules.
2. Use git filter-branch on each submodule to remove the submodule and rewrite the history so that it looks like the submodule's contents were always part of the main repository.
3. Create a mapping between the original commits in the submodules and the new commits in the main repository using git replace.
4. Modify the autolinks in the commit messages of the main repository to point to the new commits in the main repository using a script or tool.
5. Push the modified main repository and the rewritten submodules to GitHub.
